### PR TITLE
Application: FuzzLand

### PR DESCRIPTION
**Pull request by shouc**
#### Project Abstract

> Please replace these instructions with a brief description of your project summarising key points (1-2 paragraphs).
>
> If your application is a follow-up to a previous grant, please mention which one in the first line of the abstract and include a link to previous pull requests if applicable.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] A BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (aUSD/USDT) address for the payment of the milestones is provided in the application.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. This sounds like an interesting project. Regarding the application, something to be aware of is that we usually focus on the technical details. So it would be nice if you could add more details, for example, regarding your optimistic rollups. Will you use zero-knowledge proofs for it? Do you rely on a single sequencer, etc.? Also, how does the Offchain oracle work, and do you have mock-ups or designs for the UI?

**Comment by shouc** [comment]
Hello @Noc2, thanks for the comments. I have addressed them in the update. Here are the short answers to your questions:

> Will you use zero-knowledge proofs for it?

No. Optimistic rollup is enough for our business logic. However, using ZKP to share test cases can be a potential future work.

> Do you rely on a single sequencer?

Our implementation of optimistic rollup pallets will open up the possibility of having either a single sequencer or multiple validators. 

> how does the Offchain oracle work?

There are two oracles. 

* auditor nodes take a partition of a program and do dynamic analysis on it. When they find test cases leading to new coverage or new vulnerabilities, they submit a transaction to share them. 
* validator nodes generate partition graphs, verify them, and evaluate the testcase submitted by auditor nodes.

Both oracles are implemented as light clients of the chain and interact with the optimistic rollup pallets.

**Comment by keeganquigley** [comment]
Thanks @shouc for adding the information. I will mark your application as ready for review and ping the committee. In the meantime, I have a couple of follow-up questions:

- Can you specify which framework you will use for the web app? (Vue, React, etc.)
- So If I understand correctly, other projects can deploy their ink! smart contracts on your chain in order to have them audited? 
- I haven't done much studying on automated auditing solutions, is this becoming an emerging sector in Web2? Since you mentioned that so far other crypto auditing firms still rely on human auditors.
- You mentioned that you already wrote a PoC in Solidity but the gas fees were too high. I'm curious if you tested on any L2 chains with low gas fees like Moonbeam, Polygon, Avalanche, etc. to see if the concept works well?

**Comment by shouc** [comment]
Thanks!

> Can you specify which framework you will use for the web app? (Vue, React, etc.)

We are going to use Vue. A PoC has already been built here: https://github.com/fuzzland/protocol_v1/tree/main/frontend

> So If I understand correctly, other projects can deploy their ink! smart contracts on your chain in order to have them audited?

They don't need to be deployed anywhere. The FuzzLand platform just needs its bytecode to do the auditing. Project owner can submit the bytecode to FuzzLand platform through onboard transactions. Downstream DeFi that uses the auditing data (e.g., test coverage, etc.) can be deployed on our chain in the form of ink! smart contract. One possible DeFi could be giving hacking insurance for audited projects based on their test coverage.  

> I haven't done much studying on automated auditing solutions, is this becoming an emerging sector in Web2? Since you mentioned that so far other crypto auditing firms still rely on human auditors.

It is very popular in Web2. Google, for example, has applied fuzz testing to all the code they have written and used ([source](https://www.theregister.com/2022/09/08/google_fuzz_rewards/)) and found 27000 vulnerabilities in Chrome ([source](https://info.forallsecure.com/rs/112-FGI-163/images/inf-why-fuzz-test.pdf)). Most Web3 auditing firms today rely on human but some automated auditing firms have emerged recently. For instance, Certora uses formal verification, and Veridise uses static analysis to find vulnerabilities in smart contracts. 

> You mentioned that you already wrote a PoC in Solidity but the gas fees were too high. I'm curious if you tested on any L2 chains with low gas fees like Moonbeam, Polygon, Avalanche, etc. to see if the concept works well?

We evaluated Polygon and Optimism, however, their gas fees (>$5 per fraud-proof) are still too high for our needs of efficient and cheap optimistic rollups that require multiple rounds of challenging. The main challenge with EVM chains is EVM bytecode's limited expressiveness, making it difficult to support on-chain test case evaluation and partition plan verification. Additionally, some of the algorithms and tools we require are already implemented in Rust or C and can be reused in Substrate pallets, but not in EVM.

**Review by keeganquigley — DISMISSED** [review]
@shouc Nice! Thanks for the clarifications. I see you have archived the [front-end repo](https://github.com/fuzzland/protocol_v1) as read-only so I'm assuming you are planning to build this out more as part of the grant, along with the chain.

If that's the case, I think this could be a useful project for the community if successful, and I agree that the optimistic-rollup pallet could be useful and be re-used by the community, so I am happy to approve and see what you come up with.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the updates. I'm generally happy to go ahead with it. Could you still integrate some of the information that you share here into the specification of the milestones? For example, the link to the PoC of the front end and that you are going to use Vue for it. The milestone delivery tables are the requirements of our contracts, so it’s nice to have as many details there as possible.

**Review by dsm-w3f — COMMENTED** [review]
@shouc thank you for your application. In general, I think is a very interesting project. I have some questions. Are you [aware of this project](https://github.com/w3f/Grants-Program/pull/989) and [this RFP](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/decentralized-security-marketplace.md)? Could you please compare your project with them? Why do you choose one milestone (5 months) approach? Couldn't this be a little bit risky? As we pay only after the delivery and approval of the milestone, could you run into a cashflow problem? Is this project funded by any other source?

File: applications/FuzzLand.md [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+
+Fuzz testing employs partitioning through the use of an instrumented target. If an input causes execution of code outside the partition plan, the target will terminate. Early termination reduces the time spent exploring code not within the partition, saving significant time.
+
+
+**Reaching Consensus**
+
+Verifying partition plans and interesting test cases can be costly or even impossible on the chain. Thus, validator nodes use off-chain oracles. FuzzLand uses rollup techniques to move the oracle results onto the chain and reach consensus. Specifically, an optimistic rollup pallet is implemented to achieve consensus on partition plans and interesting test cases. Once a validator node mints a partition plan or an auditor node mints a test case NFT, other validator nodes can submit fraud proofs to challenge it within 50 blocks, or it will be committed. Unlike human auditors or judges, validator nodes can find evidence to challenge false claims in microseconds, as the verification process is automated and inexpensive, making optimistic rollups effective. 
+
```
Comment (bhargavbh) [inline applications/FuzzLand.md:131]: interesting application scenario for roll-ups. worthwhile to explore ideas of mechanism design to ensure faster partition plan verification, maybe in a follow-up grant. ideas like [fair-cake-cutting](https://en.wikipedia.org/wiki/Fair_cake-cutting), allowing more decisions to be made on-chain

**Review by bhargavbh — COMMENTED** [review]
Overall its an interesting project, which tries to achieve a trust-less parallelised auditing platform. It would be a cool Dapp to have on Substrate.

**Comment by randombishop** [comment]
@shouc Thank you for this application, I find this project fascinating.

I only have a couple of questions regarding the off-chain program that will actually run the LLVM and report vulnerabilities: can you please provide more details on this deliverable:
- What frameworks/libraries will it rely upon? (I am guessing you will not be developing the core of it from scratch, right?)
- Your application mentions that the vulnerability search is not solely based on random brute force parameters search; I am curious just for personal learning about how these heuristics work? AI techniques? Can you please point me to relevant documentation to learn more? 

PS: Please include infos for my first question in the application; but the second question doesn't have to be part of it, just asking for myself :)

File: applications/FuzzLand.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+ 
+In contrast, FuzzLand aims to parallelize novel automated program analysis techniques to gain accurate results in a reasonable amount of time. To achieve high parallelism with low costs, FuzzLand platform allows the public to contribute computation power to accomplish the automated auditing in return for token rewards. In the meantime, all the program analysis intermediate statistics and waypoints are verified and stored on the FuzzLand chain, which can finally be leveraged to mint the auditing reports. 
+
+Unlike traditional collaborative manual auditing platforms, FuzzLand uses sound automated program analysis techniques to provide auditing reports. Since the program analysis results and intermediate waypoints can be easily verified through a fully automated oracle, the manual confirmation process is no longer needed. While it is impossible to quantify the performance of human auditors, FuzzLand can quantify the auditing progress and completeness of auditing reports based on metrics backed with on-chain data.
```
[the remark this answers is not in the document]
Comment (shouc) [reply applications/FuzzLand.md]: I have fixed this expression in the updated version. Sound automated program analysis means that the analysis yields no false positives. Among common strategies, fuzzing is usually known as the only sound program analysis technique. FuzzLand uses fuzzing, symbolic execution, and concolic execution. Note that though symbolic execution and concolic execution are not sound, fuzzing will confirm their outcome by a concrete execution of the target.

For details, you can refer to this paper about how fuzzing and symbolic/concolic execution are integrated, yielding zero-FP results: https://sites.cs.ucsb.edu/~vigna/publications/2016_NDSS_Driller.pdf

File: applications/FuzzLand.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+3. When auditor nodes detect a test case leading to vulnerability or new code coverage, they mint an NFT for the test case. Judge nodes then verify the test case by re-executing it. Since the execution is deterministic, the outcome must also be deterministic, allowing validator nodes to reach consensus about the validity of test case NFTs. 
+
+4. Project owner can terminate an audit request. The termination requests also mint an NFT for an auditing report automatically based on the intermediate statistics, test cases, and vulnerabilities. A reward (i.e., bounties) is then given to the validator nodes and auditor nodes that generated or verified the test case NFTs.
+
+
+#### Technical Details
+
+**Partitioning Plan Synthesis**
```
[the remark this answers is not in the document]
Comment (shouc) [reply applications/FuzzLand.md:88]: cyclomatic complexity can only determine full CFG complexity (i.e., full program complexity). Here, we need edge complexity (i.e., the complexity of each branch in the program).

File: applications/FuzzLand.md [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+**Dynamic Program Analysis (DPA)**
+
+We support automated auditing of any program that can be compiled to LLVM bytecode by leveraging fuzz testing techniques, which involve sending random input to the program. This method, also known as heuristic search, aims to achieve 100% code coverage and uncover all vulnerabilities. While infinite time would guarantee zero false negatives, we use formal methods such as symbolic and concolic execution for guiding the fuzz testing search to reduce the time needed. Additionally, by partitioning the program into smaller, more manageable subprograms for each node, we can reduce the time required linearly as the number of nodes increases.
+
+Fuzz testing employs partitioning through the use of an instrumented target. If an input causes execution of code outside the partition plan, the target will terminate. Early termination reduces the time spent exploring code not within the partition, saving significant time.
+
+
```
[the remark this answers is not in the document]
Comment (shouc) [reply applications/FuzzLand.md:127]: Yes! Symbolic/concolic execution by default supports solving for a partial program (a partition in our context).

File: applications/FuzzLand.md (hunk 1 above)
Comment (shouc) [reply applications/FuzzLand.md:131]: Sounds good!

**Comment by shouc** [comment]
Thanks for the comments! I have made some changes to the proposal. 

> Are you https://github.com/w3f/Grants-Program/pull/989 and [this RFP](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/decentralized-security-marketplace.md)? Could you please compare your project with them?

I have mentioned them and added the comparisons between FuzzLand and QRUCIAL DAO in the updated proposal. 

> Why do you choose one milestone (5 months) approach? Couldn't this be a little bit risky? As we pay only after the delivery and approval of the milestone, could you run into a cashflow problem? Is this project funded by any other source?

I have added a new milestone in the proposal. We are about to finish raising our seed round so there should not be any cashflow problem for now. 

> What frameworks/libraries will it rely upon? (I am guessing you will not be developing the core of it from scratch, right?)

Yes! I have added a dependencies section in the updated proposal. 

> Your application mentions that the vulnerability search is not solely based on random brute force parameters search; I am curious just for personal learning about how these heuristics work? AI techniques? Can you please point me to relevant documentation to learn more?

Yes, it is somewhat like an AI conducting heuristic searches over the program space. Here are some great documents about fuzzing from Google: https://github.com/google/fuzzing and forallsecure: https://forallsecure.com/resources/ultimate-guide-to-fuzz-testing

**Review by randombishop — APPROVED** [review]
Thanks for the update and answers @shouc !
I am happy to approve this application.

**Review by keeganquigley — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]
I'm also happy to go ahead with it.

**Comment by sixTheDave** [comment]
@shouc Hello, I just found your proposal and saw you are building something that could be interesting for collab with Qrucial DAO. We are more focused on the reputation system, manual audits and improving the quality of them, while you are focusing on automation and payment.

Lets talk about it. You can reach me here: [https://sixthedave.me/](https://sixthedave.me/)

**Comment by keeganquigley** [comment]
Hi @shouc is it possible to provide an update on Milestone 1?

**Comment by keeganquigley** [comment]
pinging @shouc for a friendly reminder.

**Comment by keeganquigley** [comment]
Hi @shouc are you still working on the grant? If you are experiencing delays we would ask for an amendment to be signed in order to extend the timeline. I will email you as well in case you aren't seeing these comments.

**Comment by keeganquigley** [comment]
Hi @shouc please note that if I don't hear back from you after two weeks, the committee will likely cancel the grant due to inactivity. Let me know if you wish to continue work on it.

**Pull request merged**